### PR TITLE
feature: experimentalist pipeline

### PR DESCRIPTION
**Pull request by hollandjg**
## Description

Create a first pipeline for the experimentalists. 

Each pipeline has a pool method and then a sequence of filters which manipulate the results of the step before.

### Type of change:
- New feature (non-breaking change which adds functionality)

### New features:
- the `Pipeline` itself, and a `make_pipeline` factory function
- Example code and tests for creating the pipeline with custom functions
- Examples of useful pipeline elements: 
  - an exhaustive grid-search pooler, 
  - a filter, and 
  - a random sampler

Note: the "uncertainty sampler" has an additional dependency which is broken for python 3.10, so it's been moved to a separate PR #143

### CODE REVIEW

**Comment by TheLemonPig** [comment]
This all looks really nice and clean :)

**Review by gtdang — COMMENTED** [review]
This is great! A great starting point to work with. 

One thing we ought to think about is how we want to handle filters that need to interact with each other. One that comes to mind is that if a downstream validator/filter needs to resample the pool using the sampler's method. e.g. [pool, sampler, validator] I suppose this could be done in two ways. 

1.  The pipe class could save the sampler callable in the init (similar to the pool) and the validator has some means of accessing it. 
2. The validator is initialized with a sampler callable as a keyword before it is passed into the pipe.  

I'm thinking 2 is cleaner.

**Comment by TheLemonPig** [comment]
@gtdang that's a good point! yes 2 does sound cleaner

File: tests/test_experimentalist.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+    # Is sampling the number of trials we expect
+    assert len(results) == n_trials
+
+    # Filter is selecting where IV1 >= IV2
+    assert all([s[0] >= s[1] for s in results])
+
+    # Is sampling randomly
+    results2 = pipeline_random_samp.run()
+    assert results != results2
```
Comment (benwandrew) [inline tests/test_experimentalist.py:140]: could these in principle be equal, albeit with a vanishingly small probability?

Comment (gtdang) [reply tests/test_experimentalist.py:140]: Yeah that's true. I thought about that. Just a quick a dirty way to test it as we're developing! I suppose a more robust version with less the probability would be to run it many more times and make sure the results are not all the same.

Comment (benwandrew) [reply tests/test_experimentalist.py:99]: gotcha! quick and dirty to get things in motion certainly makes sense, i just couldn't resist asking :)

File: autora/experimentalist/sampler.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import random
+from typing import Iterable, Sequence, Union
+
+import numpy as np
+from alipy.query_strategy.query_labels import QueryInstanceUncertainty
```
Comment (hollandjg) [inline autora/experimentalist/sampler.py:5]: Hey @gtdang  – alipy seems broken when I run it, and I don't think we necessarily want to have it as a dependency. How easy would it be for us to do our own uncertainty sampler? It looks pretty straightforward.

Comment (hollandjg) [reply autora/experimentalist/sampler.py:5]: Looks like it was a python 3.9+ issue – ALiPy doesn't work with 3.9 or 3.10. Same thing – I think we need to think about whether we want to keep it as a dependency.

**Review by gtdang — APPROVED** [review]
Looked at PR together with John.

**Pull request merged**